sky/data/storage.py

```diff
@@ -843,47 +876,58 @@ def delete(self) -> None:
     def get_handle(self) -> StorageHandle:
         return aws.resource('s3').Bucket(self.name)
 
-    def sync_local_dir(self) -> None:
-        """Syncs a local directory to a S3 bucket.
+    def batch_aws_rsync(self,
+                        source_path_list: List[Path],
+                        create_dirs: bool = False) -> None:
+        """Invokes aws s3 sync to batch upload a list of local paths to S3
 
         AWS Sync by default uses 10 threads to upload files to the bucket.  To
         increase parallelism, modify max_concurrent_requests in your aws config
         file (Default path: ~/.aws/config).
+
+        Since aws s3 sync does not support batch operations, we construct
+        multiple commands to be run in parallel.
+
+        Args:
+            source_path_list: List of paths to local files or directories
+            create_dirs: If the local_path is a directory and this is set to
+                False, the contents of the directory are directly uploaded to
+                root of the bucket. If the local_path is a directory and this is
+                set to True, the directory is created in the bucket root and
+                contents are uploaded to it.
         """
-        source = os.path.abspath(os.path.expanduser(self.source))
-        sync_command = ('aws s3 sync --no-follow-symlinks '
-                        f'{source} s3://{self.name}/')
+
+        def get_file_sync_command(base_dir_path, file_names):
```

Comment:
`get_file_sync_command` and `get_dir_sync_command` can be united under a single function. (in that function it will check if the uploaded file/folder is a file/folder). This would make `parallel_upload` a lot simpler.

Reply:
Ah, the reason I did it this way was because that would result in code duplication in S3Store and GCStore classes. E.g., both would implement something like:

```python
def get_sync_command(source_path_list):
    commands = []
    grouped_files, dirs = data_utils.group_files_by_dir(source_path_list)
    # Generate file upload commands
    for dir_path, file_names in grouped_files.items():
        sync_command = get_file_sync_command(dir_path, file_names)
        commands.append(sync_command)
    # Generate dir upload commands
    for dir_path in dirs:
        if create_dirs:
            dest_dir_name = os.path.basename(dir_path)
        else:
            dest_dir_name = ''
        sync_command = get_dir_sync_command(dir_path, dest_dir_name)
        commands.append(sync_command)
    return commands
```

I figured it's cleaner to extract this method into data_utils.py instead of duplicating it in the two classes or implementing in a parent class. lmk if you feel strongly otherwise, I can change this

Reply:
Hmm yeah the former options seems better than. Its less LOC. Sounds good!